google-cloud-storage/src/main/java/com/google/cloud/storage/BidiAppendableUnbufferedWritableByteChannel.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import com.google.cloud.BaseServiceException;
+import com.google.cloud.storage.ChunkSegmenter.ChunkSegment;
+import com.google.cloud.storage.UnbufferedWritableByteChannelSession.UnbufferedWritableByteChannel;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class BidiAppendableUnbufferedWritableByteChannel implements UnbufferedWritableByteChannel {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(BidiAppendableUnbufferedWritableByteChannel.class);
+
+  private final BidiUploadStreamingStream stream;
+  private final ChunkSegmenter chunkSegmenter;
+
+  private boolean open;
+  private long writeOffset;
+  private volatile boolean nextWriteShouldFinalize;
+  private boolean writeCalledAtLeastOnce;
+
+  /** If write throws an error, don't attempt to finalize things when {@link #close()} is called. */
+  private boolean writeThrewError;
+
+  BidiAppendableUnbufferedWritableByteChannel(
+      BidiUploadStreamingStream stream, ChunkSegmenter chunkSegmenter, long writeOffset) {
+    this.stream = stream;
+    this.chunkSegmenter = chunkSegmenter;
+    this.open = true;
+    this.writeOffset = writeOffset;
+    this.nextWriteShouldFinalize = false;
+    this.writeThrewError = false;
+  }
+
+  @Override
+  public long write(ByteBuffer[] srcs, int srcsOffset, int srcsLength) throws IOException {
+    return internalWrite(srcs, srcsOffset, srcsLength);
+  }
+
+  @Override
+  public long writeAndClose(ByteBuffer[] srcs, int offset, int length) throws IOException {
```

Comment:
Why isn't this just

  long written = write(srcs, offset, length);
  close();
  return written;